--- a/src/fido/cbor_make_credential.c
+++ b/src/fido/cbor_make_credential.c
@@ -384,12 +384,8 @@
     }
 
     if (options.up == ptrue || options.up == NULL) { //14.1
-        if (pinUvAuthParam.present == true) {
-            if (getUserPresentFlagValue() == false) {
-                if (check_user_presence() == false) {
-                    CBOR_ERROR(CTAP2_ERR_OPERATION_DENIED);
-                }
-            }
+        if (!check_user_presence_with_pin((flags & FIDO2_AUT_FLAG_UV) != 0)) {
+            CBOR_ERROR(CTAP2_ERR_OPERATION_DENIED);
         }
         flags |= FIDO2_AUT_FLAG_UP;
         if (options.up == ptrue) {
